ldap/servers/plugins/alias_entries/alias-entries.c

```diff
@@ -0,0 +1,123 @@
+/** BEGIN COPYRIGHT BLOCK
+ * Copyright (C) 2023 anilech
+ * Copyright (C) 2023 Red Hat, Inc.
+ * All rights reserved.
+ *
+ * License: GPL (version 3 or any later version).
+ * See LICENSE for details.
+ * END COPYRIGHT BLOCK **/
+
+#include "alias-entries.h"
+
+static Slapi_ComponentId *plugin_id = NULL;
+Slapi_PluginDesc srchpdesc = { PLUGINNAME, PLUGINVNDR, PLUGINVERS, PLUGINDESC };
+
+static Slapi_DN *
+alias_get_next_dn(Slapi_DN *sdn)
+{
+    Slapi_PBlock *pb = NULL;
+    Slapi_Entry **e;
+    Slapi_DN *entry_sdn = NULL;
+    int derefNever = LDAP_DEREF_NEVER;
+    char *attrs[] = { "aliasedObjectName", NULL };
+
+    /* search dn to check if it is an alias */
+    pb = slapi_pblock_new();
+    slapi_search_internal_set_pb_ext(pb, sdn, LDAP_SCOPE_BASE, ALIASFILTER, attrs, 0, NULL, NULL, plugin_id, 0);
+    slapi_pblock_set(pb, SLAPI_SEARCH_DEREF, (void *)&derefNever);
+
+    if (slapi_search_internal_pb(pb) == 0) {
+        slapi_pblock_get(pb, SLAPI_PLUGIN_INTOP_SEARCH_ENTRIES, &e);
+        if (e && e[0]) {
+            entry_sdn = slapi_sdn_new_dn_byval(slapi_entry_attr_get_ref(e[0], attrs[0]));
+        }
+    }
+    slapi_free_search_results_internal(pb);
+    slapi_pblock_destroy(pb);
+
+    return entry_sdn;
+}
+
+int
+alias_entry_init(Slapi_PBlock *pb)
+{
+    if (slapi_pblock_set(pb, SLAPI_PLUGIN_VERSION, SLAPI_PLUGIN_VERSION_03) != 0 ||
+        slapi_pblock_set(pb, SLAPI_PLUGIN_DESCRIPTION, (void *)&srchpdesc) != 0 ||
+        slapi_pblock_set(pb, SLAPI_PLUGIN_PRE_SEARCH_FN, (void *) alias_entry_srch) != 0  ||
+        slapi_pblock_get(pb, SLAPI_PLUGIN_IDENTITY, &plugin_id) != 0)
+    {
+        slapi_log_error(SLAPI_LOG_ERR, PLUGINNAME,
+                        "alias_entry_init: Error registering the plug-in.\n");
+        return -1;
+    }
+
+    return 0;
+}
+
+int
+alias_entry_srch(Slapi_PBlock *pb)
+{
+    Slapi_DN *search_target = NULL;
+    Slapi_DN *dn1 = NULL;
+    Slapi_DN *dn2 = NULL;
+    char *str_filter = NULL;
+    int scope = 0;
+    int deref_enabled = 0;
+    size_t i = 0;
+
+    /* deref must be enabled */
+    if (slapi_pblock_get(pb, SLAPI_SEARCH_DEREF, &deref_enabled) != 0) {
+        return 0;
+    }
+    if (deref_enabled != LDAP_DEREF_FINDING && deref_enabled != LDAP_DEREF_ALWAYS) {
+        return 0;
+    }
+
+    /* Only base search is supported */
+    if (slapi_pblock_get(pb, SLAPI_SEARCH_SCOPE, &scope) != 0 || scope != LDAP_SCOPE_BASE) {
```

Comment:
@tbordaz, not sure what you were thinking exactly, but I do not think logging a message in error log is a good idea. (or at least not something logged at the default log level) because of the risk of flooding the log file. Adding a note in the access log result, or in the ldap result message may be useful.

Reply:
I agree there is a risk of flooding error logs, unless it is logged only once. Ideally the ldap_result would be the most helpful for client but the plugin does not return result.

Reply:
But the search operation does, and the plugin could set the SLAPI_PB_RESULT_TEXT in pblock ...

Reply:
> We could log a warning message that the dereferenceAlias indicator is not supported for sub/one level scope.

I decided to make it a full blown error if you don't use a base-level search.  Any concerns here?

Please review...

Reply:
Looks good !